### PR TITLE
Cleaner RVSDG translation using Phi nodes

### PR DESCRIPTION
**Pull request by oflatt**
This PR avoids extra blocks by using `phi` nodes to select between different values.
This uncovered a bug in brillvm, and also our direct jump optimizer had to take into account phi nodes.

### CODE REVIEW

**Comment by Alex-Fischman** [comment]
This PR also disables brilift in nightlies.

File: src/rvsdg/optimize_direct_jumps.rs [hunk 1]
```diff
@@ -88,6 +90,24 @@ impl SimpleCfgFunction {
...
+        for (old, new) in node_mapping.iter() {
+            assert!(
+                label_mapping
+                    .insert(
+                        self.graph[*old].name.to_string(),
+                        resulting_graph[*new].name.to_string(),
+                    )
+                    .is_none(),
+                "Duplicate labels in graph"
+            );
```
Comment (Alex-Fischman) [inline src/rvsdg/optimize_direct_jumps.rs:103]: nit: I know it's technically correct but stateful expressions in asserts is sus.

File: src/rvsdg/optimize_direct_jumps.rs [hunk 2]
```diff
@@ -99,51 +119,105 @@ impl SimpleCfgFunction {
...
+                    (
+                        self.graph[*parent].name.to_string(),
+                        self.graph[self.get_parent_skipping_empty(*parent)]
+                            .name
+                            .to_string(),
+                    )
+                })
+                .collect::<HashMap<_, _>>();
+            // if we have fewer parents, phi nodes may get messed up, so avoid the optimization
+            if new_parents.len() == parents.len() {
```
Comment (Alex-Fischman) [inline src/rvsdg/optimize_direct_jumps.rs:168]: This makes the optimization markedly less effective. If both parents become the same, that means we have a diamond pattern where the sides are empty, and we 100% want to simplify that. You should make `fixup_phis` handle the case where phi instruction become id instructions.

File: src/rvsdg/optimize_direct_jumps.rs [hunk 3]
```diff
@@ -99,51 +119,105 @@ impl SimpleCfgFunction {
...
+        if let ([source_edge], [outgoing]) = (incoming.as_slice(), outgoing.as_slice()) {
+            Some((source_edge.id(), outgoing.id()))
+        } else {
+            None
+        }
+    }
+
+    fn get_parent_skipping_empty(&self, parent_block: NodeIndex) -> NodeIndex {
+        if let Some((source_edge, _)) = self.get_single_in_single_out(parent_block) {
+            self.graph.edge_endpoints(source_edge).unwrap().0
```
Comment (Alex-Fischman) [inline src/rvsdg/optimize_direct_jumps.rs:204]: What if the grandparent is also empty? I think this function should be recursive...

**Review by Alex-Fischman — CHANGES_REQUESTED** [review]

File: src/rvsdg/optimize_direct_jumps.rs (hunk 1 above)
Comment (oflatt) [reply src/rvsdg/optimize_direct_jumps.rs:98]: fixed

File: src/rvsdg/optimize_direct_jumps.rs (hunk 2 above)
Comment (oflatt) [reply src/rvsdg/optimize_direct_jumps.rs:164]: Actually, if both parents are empty your `select` optimization should fire. So this case should be handled by egglog.

File: src/rvsdg/optimize_direct_jumps.rs (hunk 3 above)
Comment (oflatt) [reply src/rvsdg/optimize_direct_jumps.rs:204]: Changed the name of the func- it needs to not be recursive for correctness

**Comment by oflatt** [comment]
Oops, re-added branch

File: src/util.rs [hunk 4]
```diff
@@ -471,14 +466,9 @@ impl Run {
...
-                self.optimize_egglog.unwrap(),
-                self.optimize_brilift.unwrap(),
-            ) {
-                (false, false) => "-O0",
-                (true, false) => "-O0-eggcc",
-                (false, true) => "-O3",
-                (true, true) => "-O3-eggcc",
+            name += match self.optimize_brilift.unwrap() {
+                false => "-O0",
+                true => "-O3",
```
Comment (Alex-Fischman) [inline src/util.rs:471]: The brilift runmode comments that I wrote are wrong now, you should fix them.

**Review by Alex-Fischman — APPROVED** [review]
Approved once comments are fixed

**Pull request merged**